<a href="https://colab.research.google.com/github/Architag1503/Colab/blob/main/Numpy6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Numpy Standard Deviation**

In [4]:
import numpy as np
a = np.array([[1,4,7,10] , [2,5,8,11]])
b = np.std(a)
print(b)

c = np.array([[1,4,7,10] , [2,5,8,11]])
d = np.std(c , axis = 0)#along each column
print(d)
d = np.std(c , axis = 1)#along each row
print(d)

3.391164991562634
[0.5 0.5 0.5 0.5]
[3.35410197 3.35410197]


In [8]:
import numpy as np
a = np.zeros((2 , 512*512) , dtype=np.float32)
a[1 , :] = 1.0
a[0 , :] = 0.1
print(a)
b = np.std(a)
print(b)

[[0.1 0.1 0.1 ... 0.1 0.1 0.1]
 [1.  1.  1.  ... 1.  1.  1. ]]
0.45000008


# **Numpy Mean**

In [5]:
import numpy as np
a = np.array([[1,4,7,10] , [2,5,8,11]])
b = np.mean(a)
print(b)

c = np.array([[1,4,7,10] , [2,5,8,11]])
d = np.mean(c , axis = 0)#along each column
print(d)
d = np.mean(c , axis = 1)#along each row
print(d)

6.0
[ 1.5  4.5  7.5 10.5]
[5.5 6.5]


In [9]:
import numpy as np
a = np.zeros((2 , 512*512) , dtype=np.float32)
a[1 , :] = 1.0
a[0 , :] = 0.1
print(a)
b = np.mean(a)
print(b)

[[0.1 0.1 0.1 ... 0.1 0.1 0.1]
 [1.  1.  1.  ... 1.  1.  1. ]]
0.54999995


# **Numpy Argmax**

In may cases, where the size of the array is too large , it takes to much time to find maximum elements from them. For this purpose , the numpy module of python provides a function called numpy.argmax(). This function returns indices of the maximum values are returned along with specified axis.

In [14]:
import numpy as np
x = np.arange(20).reshape(4,5) + 7
print(x)
y = np.argmax(x)
z = np.argmax(x , axis = 0)
w = np.argmax(x , axis = 1)
print(y)
print(z)
print(w)

col_max_values = x[z, np.arange(x.shape[1])]
row_max_values = x[np.arange(x.shape[0]), w]

print("Column-wise max values (NumPy):", col_max_values)
print("Row-wise max values (NumPy):", row_max_values)

[[ 7  8  9 10 11]
 [12 13 14 15 16]
 [17 18 19 20 21]
 [22 23 24 25 26]]
19
[3 3 3 3 3]
[4 4 4 4]
Column-wise max values (NumPy): [22 23 24 25 26]
Row-wise max values (NumPy): [11 16 21 26]


In [20]:
import numpy as np
x = np.arange(20).reshape(5 ,4) + 10
indices = np.unravel_index(np.argmax(x, axis=None), x.shape)
print(x)
print(x[indices])

[[10 11 12 13]
 [14 15 16 17]
 [18 19 20 21]
 [22 23 24 25]
 [26 27 28 29]]
29


# **numpy.diff() in Python**

The numpy module of Python provides a function called numpy.diff for calculating the nth discrete difference along the given axis. If 'x' is the input array, then the first differnce is given by out[i] = x[i+1] - a[i]. We can calculate the higher difference by using diff recursively. The numpy module of Python provides a function called numpy.diff for calculating the nth discrete difference along axis. If 'x' is the input array , then the first differnce is given by out[i]=x[i+1]-a[i]. We can calculate the higher difference by usinf diff recursively.

In [24]:
import numpy as np
arr = np.array([0 , 1 , 2] , dtype=np.uint32)
print(arr)
b = np.diff(arr)
print(b)
print(arr[2,...] - arr[1,...] - arr[0,...])

[0 1 2]
[1 1]
1


In [25]:
import numpy as np
x = np.array([[11 , 21 , 41], [71 , 1 , 12] , [33 , 2 , 13]])
y = np.diff(x , axis=1)
print(y)
y = np.diff(x , axis=0)
print(y)

[[ 10  20]
 [-70  11]
 [-31  11]]
[[ 60 -20 -29]
 [-38   1   1]]


# **numpy.histogram() in Python**

The numpy module of Python provides a function called numpy.histogram(). This function represents the frequency of the number of values that are compared with a set of values ranges. This function is similar to the hist() function of matplotlib.pyplot

In [26]:
import numpy as np
a = np.histogram([1 , 5 , 2] , bins=[0 , 1 , 2 , 3])
print(a)

(array([0, 1, 1]), array([0, 1, 2, 3]))


In [28]:
import numpy as np
x = np.histogram(np.arange(6) , bins=np.arange(7) , density=True)
print(x)

(array([0.16666667, 0.16666667, 0.16666667, 0.16666667, 0.16666667,
       0.16666667]), array([0, 1, 2, 3, 4, 5, 6]))


Creates bin edges:

[0, 1, 2, 3, 4, 5, 6]

So the bins are:<br>
[0–1), [1–2), [2–3), [3–4), [4–5), [5–6)

np.histogram(...)<br>
np.histogram(data, bins)

Counts how many values fall into each bin.

Here:<br>
Each number (0,1,2,3,4,5) falls into a different bin

So raw counts would be:

[1, 1, 1, 1, 1, 1]

**density=True**

This is the most important part.

Instead of counts, NumPy returns a probability density:

density = count/(total samples×bin width)

Here:

Total samples = 6

Bin width = 1

So each bin value is:

1/(6×1) = 0.16666667


In [29]:
import numpy as np
a = np.arange(8)
hist , bin_edges = np.histogram(a , density=True)
print(hist)
print(bin_edges)

[0.17857143 0.17857143 0.17857143 0.         0.17857143 0.17857143
 0.         0.17857143 0.17857143 0.17857143]
[0.  0.7 1.4 2.1 2.8 3.5 4.2 4.9 5.6 6.3 7. ]


# **numpy.pad() in Python**

The numpy module of python provides called numpy.pad() to perform padding in the array. This function has several required and optional parameters.

Syntax:<br>
np.pad(array , pad_width , mode , **kwargs)


In [32]:
import numpy as np
x = [1 , 3 , 2 , 5 , 4]
y = np.pad(x , (3, 2) , 'constant')
print(y)
y = np.pad(x , (3, 2) , 'constant' , constant_values=(6 , 4))
print(y)

# pad_width = (3, 2)
# (3, 2)

# Means:
# Add 3 values to the left
# Add 2 values to the right

# mode='constant'
# This tells NumPy:
# “Pad using constant values (not reflection, edge, etc.)”

# constant_values = (6, 4)
# Means:
# Use 6 for left padding
# Use 4 for right padding

[0 0 0 1 3 2 5 4 0 0]
[6 6 6 1 3 2 5 4 4 4]


In [31]:
import numpy as np
x = [1 , 3 , 2 , 5 , 4]
y = np.pad(x , (3 , 2) , 'edge')
print(y)

#Extend the values of left and right edges number of times that is (3 , 2) => (1 1 1 , 4 4)

[1 1 1 1 3 2 5 4 4 4]


In [34]:
import numpy as np
x = [1 , 3 , 2 , 5 , 4]
y = np.pad(x , (3,2) , 'linear_ramp' , end_values=(-4 , 5))
print(y)

#So linear_ramp will continously or linearly decreasing the value from left and
#increasing the value from right for given number of times that is (3 , 2)

[-4 -3 -1  1  3  2  5  4  4  5]


In [37]:
import numpy as np
x = [1 , 3 , 2 , 5 , 4]
# mean , median , minimum , maximum
y = np.pad(x , (3,2) , 'maximum')
print(y)
y = np.pad(x , (3,2) , 'median')
print(y)
y = np.pad(x , (3,2) , 'mean')
print(y)

[5 5 5 1 3 2 5 4 5 5]
[3 3 3 1 3 2 5 4 3 3]
[3 3 3 1 3 2 5 4 3 3]


# **numpy.ravel() in Python**


The numpy module of python provides a function called numpy.ravel() , which is used to change the 2-dimensional array or multi-dimensional array into contigous flattened array. The returned array has the same data type as the source array or input array. If the input array is a masked array , the returned array will also be a masked array.

In [39]:
import numpy as np
x = np.array([[1 , 3 , 5] ,
              [11 , 35 , 56]])
y = np.ravel(x)
print(y)
y = np.ravel(x , order='F')
z = np.ravel(x , order='C')
p = np.ravel(x , order='A')
q = np.ravel(x , order='K')
print(y)
print(z)
print(p)
print(q)

[ 1  3  5 11 35 56]
[ 1 11  3 35  5 56]
[ 1  3  5 11 35 56]
[ 1  3  5 11 35 56]
[ 1  3  5 11 35 56]


In [40]:
import numpy as np
x = np.arange(12).reshape(3 , 2 , 2).swapaxes(1 , 2)
print(x)

y = np.ravel(a, order='C')
print(y)

z = np.ravel(x , order='K')
print(z)

q = np.ravel(x , order='F')
print(q)

[[[ 0  2]
  [ 1  3]]

 [[ 4  6]
  [ 5  7]]

 [[ 8 10]
  [ 9 11]]]
[0 1 2 3 4 5 6 7]
[ 0  1  2  3  4  5  6  7  8  9 10 11]
[ 0  4  8  1  5  9  2  6 10  3  7 11]


# **numpy.save() in Python**

The numpy module of python provides a function called numpy.save() to save an array into binary file in .npy format. In may of the caes, we require data in binary format to manipulate it.

In [43]:
import numpy as np
from tempfile import TemporaryFile
out_file = TemporaryFile()
x = np.arange(15)
print(x)
np.save(out_file, x)
out_file.seek(0)#Only needed here to simulate closing and reopening file
np.load(out_file)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14]


array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14])

In [44]:
import numpy as np
from tempfile import TemporaryFile
out_file = TemporaryFile()
x = np.arange(15)
print(x)
np.save(out_file, x , allow_pickle=False)
out_file.seek(0)#Only needed here to simulate closing and reopening file
np.load(out_file)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14]


array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14])